### ISSUE TITLE
useDialog: Pinch-to-zoom on iOS

### ISSUE THREAD

**Issue by backflip** [adobe/react-spectrum#1898]
# 🐛 Bug Report 

I'm unable to use the pinch-to-zoom gesture to zoom in or out when a dialog is opened. 

## 🤔 Expected Behavior

The default browser behavior of zooming in and out should work within a dialog as the content might become inaccessible otherwise.

## 😯 Current Behavior

On the [demo page](https://react-spectrum.adobe.com/react-aria/useDialog.html), focussing a field inside the dialog on my iPhone will automatically zoom in (as the input font size is [below 16px](https://css-tricks.com/16px-or-larger-text-prevents-ios-form-zoom/)) but I'm unable to zoom out. When starting the gesture outside of the dialog, it will close, when starting within, nothing happens.

1. Before opening:
![IMG_1266](https://user-images.githubusercontent.com/654171/117584462-7b85f580-b10d-11eb-9757-2cafc81b2de5.png)

2. After opening and focussing a field:
![IMG_1267](https://user-images.githubusercontent.com/654171/117584461-7b85f580-b10d-11eb-8f30-7d135f6cc1b4.png)

3. After attempting to reduce the zoom level by starting pinch-to-zoom outside of the dialog:
![IMG_1268](https://user-images.githubusercontent.com/654171/117584456-79bc3200-b10d-11eb-9dcd-997ba14d1432.png)

## 💁 Possible Solution

I would assume that some touch event listeners are interfering with the default behavior.

## 🔦 Context

The dialog context might become inaccessible either by cutting it off when auto-zooming or by simply being to small to be readable (where default pinch-to-zoom behavior would solve the problem).

## 💻 Code Sample

https://react-spectrum.adobe.com/react-aria/useDialog.html

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | Current version on the official demo page
| Browser          | Latest Safari on iOS 14.4.2
| Operating System | iOS 14.4.2

**Comment by LFDanLu** [comment]
Did some digging, looks like https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/overlays/src/usePreventScroll.ts#L50 is blocking pinch zooming. Will have to dig into this code to see if there is a way to resolve this without removing the background window scroll blocking.

**Comment by snowystinger** [comment]
Maybe it will include the ctrl key as noted in this function https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/interactions/src/useScrollWheel.ts#L25

**Comment by backflip** [comment]
> Did some digging, looks like https://github.com/adobe/react-spectrum/blob/main/packages/@react-aria/overlays/src/usePreventScroll.ts#L50 is blocking pinch zooming. Will have to dig into this code to see if there is a way to resolve this without removing the background window scroll blocking.

Oh my, this looks rather tricky. Thanks for digging into it!

**Comment by devongovett** [comment]
I think we can fix the issue in the docs by making the font size on the input elements larger so it doesn't zoom. As for zoom, that will be very challenging.

**Comment by devongovett** [comment]
It should be noted that zooming is still possible via the buttons in the toolbar, which is actually more similar to how desktop browsers behave anyway. Rather than cropping the viewport, it actually affects the layout of the app.

![image](https://user-images.githubusercontent.com/19409/118194191-d9f8ef80-b3fd-11eb-9de8-39bce8b8f19e.png)

**Comment by backflip** [comment]
> It should be noted that zooming is still possible via the buttons in the toolbar

Good point, did not think of that. Do you know whether similar controls exist in "WebKit wrappers" like Chrome on iOS, too?

**Comment by Haakam21** [comment]
Any progress on this? Still an issue to this day

**Comment by DanielSpindler** [comment]
Any Progress?

**Comment by attitude** [comment]
Hi, I’m testing different experiences by patching `usePreventScroll` from `@react-aria/overlays@^3.22.0”`.

**Issue:** Unable to restore zoom to 100%, it happens when:

1. Using
   - interact with any input and iOS zooms to inputs on user interaction
   - zooms page in manually
1. Users trigger a modal dialog without reverting the zoom
1. The dialog opens, and the UI is clipped
1. It’s impossible to correct the zoom

Our inputs must have a font size of 14px due to historical reasons. We want to increase the font size to reduce issues caused by Safari auto-zoom. The issue persists when users pinch to zoom the page themselves. They get stuck in a cropped UI with limited options to restore the original state.

**Have you considered adding a condition like `window.visualViewport.scale > 1` to allow gestures?**

Preventing scrolling is important for other reasons. This way, the hook can restore its original functionality after zooming out, which could be a good enough compromise.

It could be further limited to touches with a condition like `e.touches.length === 2` (or more robust tests for gesture identity or type) to restrict it further if needed.

Alternatively, users could opt out of it completely with a prop.

## Example patch limited to 2 touches & scale > 1

I’ve tested on both the iPhone 15 Pro Max (Simulator) and iPhone 16 Pro (real device) with iOS 18.1 and 18.3, respectively. I’ve added the same logic to all three listeners. From logs, only `onTouchStart` and `onTouchMove` seem to be needed.

- It allows restoring the scale back to 1.0 using two touches (pinch to zoom)
- It uses the original scroll prevention behavior after restoring the scale

I can open a new PR for this, but I saw an old one. I need to know the criteria for an accepted solution and any other potential blockers I might be unaware of.

Either way, thanks for these great tools. I hope to hear from you soon.

```diff
diff --git a/dist/usePreventScroll.mjs b/dist/usePreventScroll.mjs
index 532a72f9e9ba985d036b7eba0f950ff6552898d9..3a21c8d753f169e4c29085c1fd00ee0dcaf2377b 100644
--- a/dist/usePreventScroll.mjs
+++ b/dist/usePreventScroll.mjs
@@ -81,6 +81,9 @@ function $49c51c25361d4cd2$var$preventScrollMobileSafari() {
     let scrollable;
     let restoreScrollableStyles;
     let onTouchStart = (e)=>{
+        if (e.touches.length === 2 && window.visualViewport.scale > 1) {
+            return;
+        }
         // Store the nearest scrollable parent element from the element that the user touched.
         scrollable = (0, $7mMvr$getScrollParent)(e.target, true);
         if (scrollable === document.documentElement && scrollable === document.body) return;
@@ -90,6 +93,9 @@ function $49c51c25361d4cd2$var$preventScrollMobileSafari() {
         if (scrollable instanceof HTMLElement && window.getComputedStyle(scrollable).overscrollBehavior === "auto") restoreScrollableStyles = $49c51c25361d4cd2$var$setStyle(scrollable, "overscrollBehavior", "contain");
     };
     let onTouchMove = (e)=>{
+        if (e.touches.length === 2 && window.visualViewport.scale > 1) {
+            return;
+        }
         // Prevent scrolling the window.
         if (!scrollable || scrollable === document.documentElement || scrollable === document.body) {
             e.preventDefault();
@@ -104,6 +110,9 @@ function $49c51c25361d4cd2$var$preventScrollMobileSafari() {
         if (scrollable.scrollHeight === scrollable.clientHeight && scrollable.scrollWidth === scrollable.clientWidth) e.preventDefault();
     };
     let onTouchEnd = (e)=>{
+        if (e.touches.length === 2 && window.visualViewport.scale > 1) {
+            return;
+        }
         let target = e.target;
         // Apply this change if we're not already focused on the target element
         if ($49c51c25361d4cd2$var$willOpenKeyboard(target) && target !== document.activeElement) {
```

**Comment by shawnmclean** [comment]
Any workaround to have a modal showing image that we can zoom in on?
